modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/metadata/IgniteMdColumnOrigins.java

```diff
@@ -0,0 +1,325 @@
+            }
+        };
+        final List<RexNode> projects = new ArrayList<>();
+
+        for (RexNode rex: rexShuttle.apply(rel.getProgram().getProjectList()))
+            projects.add(rex);
+
+        final RexNode rexNode = projects.get(iOutputColumn);
+
+        if (rexNode instanceof RexInputRef) {
+            // Direct reference:  no derivation added.
+            RexInputRef inputRef = (RexInputRef) rexNode;
+            return mq.getColumnOrigins(input, inputRef.getIndex());
+        }
+        // Anything else is a derivation, possibly from multiple columns.
+        final Set<RelColumnOrigin> set = getMultipleColumns(rexNode, input, mq);
+
+        return createDerivedColumnOrigins(set);
+    }
+
+    public @Nullable Set<RelColumnOrigin> getColumnOrigins(Filter rel, RelMetadataQuery mq, int iOutputColumn) {
+        return mq.getColumnOrigins(rel.getInput(), iOutputColumn);
+    }
+
+    public @Nullable Set<RelColumnOrigin> getColumnOrigins(Sort rel, RelMetadataQuery mq, int iOutputColumn) {
+        return mq.getColumnOrigins(rel.getInput(), iOutputColumn);
+    }
+
+    public @Nullable Set<RelColumnOrigin> getColumnOrigins(TableModify rel, RelMetadataQuery mq, int iOutputColumn) {
+        return mq.getColumnOrigins(rel.getInput(), iOutputColumn);
+    }
+
+    public @Nullable Set<RelColumnOrigin> getColumnOrigins(Exchange rel, RelMetadataQuery mq, int iOutputColumn) {
+        return mq.getColumnOrigins(rel.getInput(), iOutputColumn);
+    }
+
+    public @Nullable Set<RelColumnOrigin> getColumnOrigins(TableFunctionScan rel,
+        RelMetadataQuery mq, int iOutputColumn) {
+        final Set<RelColumnOrigin> set = new HashSet<>();
+        Set<RelColumnMapping> mappings = rel.getColumnMappings();
+
+        if (mappings == null) {
+            if (!rel.getInputs().isEmpty()) {
+                // This is a non-leaf transformation:  say we don't
+                // know about origins, because there are probably
+                // columns below.
+                return null;
+            }
+            else {
+                // This is a leaf transformation: say there are fer sure no
+                // column origins.
+                return set;
+            }
+        }
+        for (RelColumnMapping mapping : mappings) {
+            if (mapping.iOutputColumn != iOutputColumn)
+                continue;
+
+            final RelNode input = rel.getInputs().get(mapping.iInputRel);
+            final int column = mapping.iInputColumn;
+            Set<RelColumnOrigin> origins = mq.getColumnOrigins(input, column);
+
+            if (origins == null)
+                return null;
+
+            if (mapping.derived)
+                origins = createDerivedColumnOrigins(origins);
+
+            set.addAll(origins);
+        }
+        return set;
+    }
+
+    public @Nullable Set<RelColumnOrigin> getColumnOrigins(IgniteIndexScan rel, RelMetadataQuery mq, int iOutputColumn) {
```

Comment:
Does it make any sense to replace this two methods with one for ProjectableFilterableTableScan